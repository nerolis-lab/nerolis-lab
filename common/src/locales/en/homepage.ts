export default {
  title: "Neroli's Lab",
  subtitle: 'Helping you overthink sleep tracking.',
  description: 'Optimize your strategies with our simulations.',
  getStarted: 'Get started',
  features: {
    title: 'Features',
    description: 'Optimize your strategies with our simulations.',
    getStarted: 'Get started'
  }
};
