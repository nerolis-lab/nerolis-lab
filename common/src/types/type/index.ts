export * from './type-paths';
export * from './path-value';
