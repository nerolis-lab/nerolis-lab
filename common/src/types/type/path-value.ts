/**
 * Utility types for extracting the value type at a given path in an object
 */

/**
 * Extract the type at a given path in an object
 * Supports dot notation, array indices, and wildcards
 *
 * @example
 * type Obj = { a: { b: { c: number } }, d: string[] };
 * type Value1 = PathValue<Obj, 'a.b.c'>; // number
 * type Value2 = PathValue<Obj, 'd.0'>; // string
 * type Value3 = PathValue<Obj, 'd.*'>; // string
 */
export type PathValue<T, Path extends string> = Path extends `${infer Head}.${infer Rest}`
  ? Head extends keyof T
    ? T[Head] extends Array<infer U>
      ? Rest extends `${number}.${infer RestAfterIndex}` | `*.${infer RestAfterIndex}`
        ? PathValue<U, RestAfterIndex>
        : Rest extends `${number}` | '*'
          ? U
          : never
      : PathValue<T[Head], Rest>
    : never
  : Path extends keyof T
    ? T[Path]
    : Path extends `${number}` | '*'
      ? T extends Array<infer U>
        ? U
        : never
      : never;
