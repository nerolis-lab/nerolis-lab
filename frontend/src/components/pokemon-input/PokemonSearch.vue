<template>
  <v-card class="py-3">
    <!-- Title, search and close button -->
    <v-row no-gutters class="flex-nowrap align-center">
      <v-col class="flex-grow-1 flex-shrink-1" style="min-width: 0">
        <v-card-title class="px-2 text-truncate">Select a Pokémon</v-card-title>
      </v-col>
      <v-col class="flex-grow-0 flex-shrink-0">
        <CustomSearchBar
          v-model="searchQuery"
          @enter="selectFirstOption"
          label="Search Pokémon, ingredient or berry"
          :start-minimized="true"
          density="compact"
          :min-width="100"
          :max-width="200"
        />
      </v-col>
      <v-col cols="auto" class="flex-shrink-0">
        <v-btn icon="mdi-close" @click="closePokemonSearch" elevation="0" />
      </v-col>
    </v-row>

    <v-row dense class="flex-left px-3 mt-0 mb-1 justify-space-between">
      <v-chip-group v-model="selectedSpecialties" multiple>
        <CustomChip
          v-for="specialty in specialties"
          :key="specialty"
          :value="specialty"
          :color="specialty"
          :is-selected="selectedSpecialties.includes(specialty)"
          :text="capitalize(specialty)"
          size="x-small"
        />
      </v-chip-group>
      <v-badge
        v-model="pokemonSearchStore.showPokeboxBadge"
        dot
        location="top right"
        color="primary"
        @click="showPokeboxInfo"
      >
        <v-btn
          @click="pokemonSearchStore.togglePokebox()"
          :title="pokemonSearchStore.showPokebox ? 'Showing Pokebox' : 'Showing All Pokemon'"
          size="small"
          elevation="0"
          icon
        >
          <v-img
            src="/images/misc/pokebox.png"
            width="36"
            height="36"
            :style="{ filter: pokemonSearchStore.showPokebox ? 'none' : 'grayscale(100%)' }"
          />
        </v-btn>
      </v-badge>
    </v-row>

    <v-row dense class="flex-center px-3 mt-0 flex-nowrap">
      <v-checkbox
        v-if="!pokemonSearchStore.showPokebox"
        v-model="finalStageOnly"
        label="Final Stage Only"
        density="compact"
        hide-details
      />
      <v-spacer />
      <DropdownSort
        v-model="selectedSort"
        v-model:sort-ascending="sortAscending"
        :sort-options="sortOptions"
        color="secondary"
      />
    </v-row>

    <v-divider class="my-2" />

    <!-- Pokemon list -->
    <div ref="pokemonListContainer" :style="containerStyle">
      <div v-if="loading" class="d-flex justify-center align-center" style="height: 200px">
        <v-progress-circular indeterminate color="primary" size="48"></v-progress-circular>
      </div>
      <div
        v-else-if="pokemonSearchStore.showPokebox && savedPokemon.length === 0"
        class="d-flex justify-center align-center pa-6"
      >
        <div class="text-center">
          <v-icon icon="mdi-pokeball" size="48" color="grey-lighten-1" class="mb-3"></v-icon>
          <div class="text-h6 text-grey-darken-1 mb-2">Your Pokebox is empty</div>
          <div class="text-body-2 text-grey-darken-2" style="max-width: 300px">
            <span v-if="!userStore.loggedIn"> Please log in to save Pokémon to your Pokebox </span>
            <span v-else> Add Pokémon from your teams to start building your personal collection </span>
          </div>
        </div>
      </div>
      <v-row v-else dense class="d-flex justify-space-around mx-1">
        <v-col v-for="{ path, pokemon, instance } in filteredPokemon" :key="instance.externalId" class="flex-center">
          <div class="flex-column align-center">
            <div v-if="pokemonSearchStore.showPokebox && instance" class="text-center text-caption mb-1">
              <div>Level {{ instance.level }}</div>
            </div>
            <v-avatar
              color="secondary"
              class="cursor-pointer"
              @click="selectPokemon(instance)"
              :size="isMobile ? 60 : 100"
              rounded="lg"
            >
              <v-img :src="path"></v-img>
            </v-avatar>
            <div v-if="pokemonSearchStore.showPokebox" class="text-center text-caption mt-1" style="max-width: 100px">
              {{ instance ? instance.name : pokemon.displayName }}
            </div>
          </div>
        </v-col>
      </v-row>
    </div>
  </v-card>

  <!-- Pokebox info dialog -->
  <v-dialog v-model="showPokeboxInfoDialog" max-width="400px">
    <v-card class="pa-4">
      <v-card-title class="text-h6 d-flex align-center">
        <v-icon icon="mdi-pokeball" color="primary" class="mr-2"></v-icon>
        What is the Pokébox?
      </v-card-title>
      <v-card-text class="text-body-2">
        <p class="mb-3">
          The Pokébox is your personal collection of Pokémon that you save to reuse them in the future.
        </p>
        <p class="mb-3">
          When you add a Pokémon to Neroli's Lab, you have the option to save it to your Pokébox either by clicking the
          bookmark icon in the top right of the Pokémon edit menu, or by clicking the Pokémon slot in the team after you
          have added it and checking "Save to Pokébox".
        </p>
        <p class="mb-0">
          Toggle between viewing all Pokémon or just your personal collection using the Pokébox button.
        </p>
      </v-card-text>
      <v-card-actions class="justify-end">
        <v-btn color="primary" @click="closePokeboxInfo">Got it!</v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script setup lang="ts">
import CustomChip from '@/components/custom-components/custom-chip/CustomChip.vue'
import DropdownSort from '@/components/custom-components/dropdown-sort/DropdownSort.vue'
import CustomSearchBar from '@/components/custom-components/search-bar/CustomSearchBar.vue'
import { useBreakpoint } from '@/composables/use-breakpoint/use-breakpoint'
import { UserService } from '@/services/user/user-service'
import { avatarImage } from '@/services/utils/image-utils'
import { PokemonInstanceUtils } from '@/services/utils/pokemon-instance-utils'
import { useDialogStore } from '@/stores/dialog-store/dialog-store'
import { usePokemonSearchStore } from '@/stores/pokemon-search-store'
import { useUserStore } from '@/stores/user-store'
import {
  capitalize,
  COMPLETE_POKEDEX,
  type Pokemon,
  type PokemonInstanceExt,
  type PokemonSpecialty
} from 'sleepapi-common'
import { computed, ref, watch, type ComputedRef, type Ref } from 'vue'

const emit = defineEmits<{
  cancel: []
  save: [pokemon: PokemonInstanceExt]
}>()

export interface PokemonWithPath {
  path: string
  pokemon: Pokemon
  instance: PokemonInstanceExt
}

const pokemonSearchStore = usePokemonSearchStore()
const userStore = useUserStore()
const dialogStore = useDialogStore()

const searchQuery = ref('')
const selectedSpecialties = ref<PokemonSpecialty[]>([])
const specialties: PokemonSpecialty[] = ['berry', 'ingredient', 'skill']
const finalStageOnly = ref(true)
const selectedSort = ref('pokedex')
const sortAscending = ref(true)
const showPokeboxInfoDialog = ref(false)

const sortOptions = computed(() => {
  const baseSortOptions = [
    { title: '#', value: 'pokedex', description: 'Sort by Pokédex number' },
    { title: 'A-Z', value: 'name', description: 'Sort alphabetically by name' }
  ]

  if (pokemonSearchStore.showPokebox) {
    return [
      ...baseSortOptions,
      { title: 'Level', value: 'level', description: 'Sort by Pokémon level' },
      { title: 'RP', value: 'rp', description: 'Sort by research power (RP)' }
    ]
  }

  return baseSortOptions
})

const { isMobile } = useBreakpoint()
const pokemonListContainer = ref<HTMLElement | null>(null)
const loading = ref(false)

// Computed style for consistent container height
// This is a pretty hacky solution, but I can't figure out a good way to not have the UI jump around
const containerStyle = computed(() => {
  const dialogMargins = 48 // Vuetify dialog margins (24px top + 24px bottom)
  const cardPadding = 24 // py-3 = 12px top + 12px bottom
  const titleRow = 40 // Title row height
  const filtersRow = 40 // Specialty chips and pokebox button row
  const checkboxRow = 48 // Final stage only and sort controls
  const divider = 12 // v-divider my-2
  const spacing = 5

  const totalFixedHeight = dialogMargins + cardPadding + titleRow + filtersRow + checkboxRow + spacing + divider

  return {
    minHeight: `calc(100vh - ${totalFixedHeight}px)`,
    transition: 'min-height 0.3s ease'
  }
})

const savedPokemon: Ref<PokemonWithPath[]> = ref([])
const completePokedex: Ref<PokemonWithPath[]> = ref(
  COMPLETE_POKEDEX.map((p) => ({
    pokemon: p,
    instance: PokemonInstanceUtils.createDefaultPokemonInstance(p),
    path: avatarImage({ pokemonName: p.name, shiny: false, happy: false })
  })).sort((a, b) => a.pokemon.pokedexNumber - b.pokemon.pokedexNumber)
)

const pokemonCollection: ComputedRef<PokemonWithPath[]> = computed(() => {
  return pokemonSearchStore.showPokebox ? savedPokemon.value : completePokedex.value
})

const filteredPokemon: ComputedRef<PokemonWithPath[]> = computed(() => {
  const query = searchQuery.value.toLowerCase().trim()

  const nameFilter = (p: PokemonWithPath) => !query || p.pokemon.displayName.toLowerCase().includes(query)
  const specialtyFilter = (p: PokemonWithPath) =>
    selectedSpecialties.value.length === 0 || selectedSpecialties.value.some((s) => p.pokemon.specialty === s)
  const finalStageFilter = (p: PokemonWithPath) =>
    !finalStageOnly.value || pokemonSearchStore.showPokebox || p.pokemon.remainingEvolutions === 0

  const filtered = pokemonCollection.value.filter(nameFilter).filter(specialtyFilter).filter(finalStageFilter)

  // Sort the filtered results
  const sorted = filtered.sort((a, b) => {
    let compareValue = 0

    switch (selectedSort.value) {
      case 'pokedex':
        compareValue = a.pokemon.pokedexNumber - b.pokemon.pokedexNumber
        break
      case 'name':
        compareValue = a.pokemon.displayName.localeCompare(b.pokemon.displayName)
        break
      case 'specialty':
        compareValue = a.pokemon.specialty.localeCompare(b.pokemon.specialty)
        break
      case 'evolution':
        compareValue = a.pokemon.remainingEvolutions - b.pokemon.remainingEvolutions
        break
      case 'level':
        compareValue = (a.instance?.level ?? 0) - (b.instance?.level ?? 0)
        break
      case 'rp':
        compareValue = (a.instance?.rp ?? 0) - (b.instance?.rp ?? 0)
        break
      default:
        compareValue = a.pokemon.pokedexNumber - b.pokemon.pokedexNumber
    }

    return sortAscending.value ? compareValue : -compareValue
  })

  return sorted
})

const selectFirstOption = () => {
  const firstOption = filteredPokemon.value[0]
  if (firstOption) {
    selectPokemon(firstOption.instance)
  }
}

const selectPokemon = (instance: PokemonInstanceExt) => {
  if (pokemonSearchStore.showPokebox) {
    // For Pokébox Pokémon, emit directly
    dialogStore.handlePokemonSelected(instance)
    emit('save', instance)
  } else {
    // For Pokédex Pokémon, open the PokemonInputDialog first
    dialogStore.openPokemonInput((updatedInstance: PokemonInstanceExt) => {
      dialogStore.handlePokemonSelected(updatedInstance)
      emit('save', updatedInstance)
    }, instance)
  }
}

const closePokemonSearch = () => {
  emit('cancel')
}

const loadPokebox = async () => {
  if (loading.value) {
    logger.debug('Already loading, skipping concurrent call')
    return
  }

  loading.value = true

  savedPokemon.value = await UserService.getUserPokemon().then((instances) =>
    instances.map((instance) => ({
      pokemon: instance.pokemon,
      instance,
      path: avatarImage({ pokemonName: instance.pokemon.name, shiny: instance.shiny, happy: false })
    }))
  )
  loading.value = false
}

const showPokeboxInfo = () => {
  showPokeboxInfoDialog.value = pokemonSearchStore.showPokeboxBadge
}

const closePokeboxInfo = () => {
  showPokeboxInfoDialog.value = false
  pokemonSearchStore.hidePokeboxBadge()
}

if (userStore.loggedIn) {
  loadPokebox()
}

// Also watch for pokebox toggle and load data if needed
watch(
  () => [pokemonSearchStore.showPokebox, userStore.loggedIn],
  async ([showPokebox, loggedIn]) => {
    if (showPokebox && loggedIn && savedPokemon.value.length === 0) {
      await loadPokebox()
    }
  },
  { immediate: true }
)
</script>
