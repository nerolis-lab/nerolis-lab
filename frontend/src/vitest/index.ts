export * from './setup'

import { commonMocks } from 'sleepapi-common'
import * as frontendMocks from './mocks/index.js'

export const mocks = {
  ...commonMocks,
  ...frontendMocks
}
