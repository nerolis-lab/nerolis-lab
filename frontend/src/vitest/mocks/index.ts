export * from './calculator'
export * from './member-iv'
export * from './pokemon-instance'
export * from './team-data'
