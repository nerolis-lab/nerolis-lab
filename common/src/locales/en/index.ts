import { default as homepage } from './files/homepage';
import { default as profilepage } from './files/profilepage';
import { default as nav } from './files/nav';

export const en = {
  homepage,
  profilepage,
  nav,

  common: {
    app: {
      name: "Neroli's Lab",
      tagline: 'Sleep Team Optimizer'
    },
    actions: {
      save: 'Save',
      cancel: 'Cancel',
      delete: 'Delete',
      edit: 'Edit',
      add: 'Add',
      remove: 'Remove',
      search: 'Search',
      filter: 'Filter',
      reset: 'Reset',
      close: 'Close',
      back: 'Back',
      next: 'Next'
    },
    status: {
      loading: 'Loading...',
      saving: 'Saving...',
      success: 'Success',
      error: 'Error',
      warning: 'Warning'
    },
    language: {
      title: 'Language'
    }
  }
};

export type Translation = typeof en;
