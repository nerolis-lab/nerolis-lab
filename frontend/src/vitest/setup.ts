import router from '@/router/router'
import { config } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import ResizeObserver from 'resize-observer-polyfill'
import { beforeEach } from 'vitest'
import 'vitest-canvas-mock'
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'
import { VTimePicker } from 'vuetify/labs/VTimePicker'
import 'vuetify/styles'

const vuetify = createVuetify({
  components: { ...components, VTimePicker },
  directives
})

config.global.plugins.push(vuetify, router)

config.global.stubs = {
  RouterView: true,
  RouterLink: true
}

vi.stubGlobal('visualViewport', new EventTarget())

if (typeof window === 'undefined') {
  vi.stubGlobal('window', {
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    setTimeout: vi.fn(),
    clearTimeout: vi.fn(),
    setInterval: vi.fn(),
    clearInterval: vi.fn(),
    location: { href: 'http://localhost:3000/' }
  })
}

global.ResizeObserver = ResizeObserver

beforeEach(() => {
  setActivePinia(createPinia())
})

afterEach(() => {
  vi.clearAllTimers()
  vi.clearAllMocks()
})
