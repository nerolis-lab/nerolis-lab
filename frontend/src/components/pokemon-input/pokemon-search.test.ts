import PokemonSearch from '@/components/pokemon-input/pokemon-search.vue'
import { faker } from '@faker-js/faker/locale/en'
import type { VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { PIKACHU, RandomUtils, uuid } from 'sleepapi-common'
import { vimic } from 'vimic'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { nextTick } from 'vue'

vi.mock('@/stores/pokedex-store/pokedex-store', () => ({
  usePokedexStore: () => ({
    groupedPokedex: [
      {
        category: 'ingredient',
        list: ['Bulbasaur', 'Charmander', 'Squirtle']
      },
      {
        category: 'berry',
        list: ['Pikachu', 'Raichu']
      },
      {
        category: 'skill',
        list: ['Machop', 'Machoke', 'Machamp']
      }
    ]
  })
}))

describe('PokemonSearch', () => {
  let wrapper: VueWrapper<InstanceType<typeof PokemonSearch>>

  beforeEach(() => {
    setActivePinia(createPinia())
    wrapper = mount(PokemonSearch, {
      props: {
        memberIndex: 0
      }
    })

    vimic(uuid, 'v4', () => 'some uuid')
    vimic(RandomUtils, 'roll', () => true)
    vimic(faker.person, 'firstName', () => 'Some name')
  })

  it('renders GroupList when no Pokémon is selected', () => {
    expect(wrapper.findComponent({ name: 'GroupList' }).exists()).toBe(true)
    expect(wrapper.findComponent({ name: 'PokemonInput' }).exists()).toBe(false)
  })

  it('emits cancel event when closeMenu is called', async () => {
    wrapper.vm.closeMenu()
    await nextTick()

    expect(wrapper.emitted('cancel')).toBeTruthy()
  })

  it('selects Pokémon correctly', async () => {
    const pkmn = PIKACHU
    wrapper.vm.selectPokemon(pkmn.name)

    await nextTick()

    expect(wrapper.findComponent({ name: 'PokemonInput' }).exists()).toBe(true)
    expect(wrapper.vm.pokemonInstance).toMatchSnapshot()
    expect(wrapper.findComponent({ name: 'PokemonInput' }).props('preSelectedPokemonInstance')).toMatchSnapshot()
  })

  it('displays error message if Pokémon is not found', async () => {
    const consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => {})
    wrapper.vm.selectPokemon('unknown')

    await nextTick()

    expect(consoleErrorSpy).toHaveBeenCalledWith('Error selecting Pokémon')
    consoleErrorSpy.mockRestore()
  })
})
