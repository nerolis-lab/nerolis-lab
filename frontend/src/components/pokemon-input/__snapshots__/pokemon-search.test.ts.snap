// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonSearch > selects Pokémon correctly 1`] = `
{
  "carrySize": 22,
  "externalId": "some uuid",
  "gender": "male",
  "ingredients": [
    {
      "amount": 1,
      "ingredient": {
        "longName": "Fancy Apple",
        "name": "Apple",
        "taxedValue": 23.7,
        "value": 90,
      },
      "level": 0,
    },
    {
      "amount": 2,
      "ingredient": {
        "longName": "Fancy Apple",
        "name": "Apple",
        "taxedValue": 23.7,
        "value": 90,
      },
      "level": 30,
    },
    {
      "amount": 4,
      "ingredient": {
        "longName": "Fancy Apple",
        "name": "Apple",
        "taxedValue": 23.7,
        "value": 90,
      },
      "level": 60,
    },
  ],
  "level": 60,
  "name": "Some name",
  "nature": {
    "energy": 1,
    "exp": 1,
    "frequency": 1,
    "ingredient": 1,
    "name": "Bashful",
    "negativeModifier": "neutral",
    "positiveModifier": "neutral",
    "prettyName": "Bashful (neutral)",
    "skill": 1,
  },
  "pokemon": {
    "berry": {
      "name": "GREPA",
      "type": "electric",
      "value": 25,
    },
    "carrySize": 17,
    "displayName": "Pikachu",
    "frequency": 2700,
    "genders": {
      "female": 0.5,
      "male": 0.5,
    },
    "ingredient0": [
      {
        "amount": 1,
        "ingredient": {
          "longName": "Fancy Apple",
          "name": "Apple",
          "taxedValue": 23.7,
          "value": 90,
        },
      },
    ],
    "ingredient30": [
      {
        "amount": 2,
        "ingredient": {
          "longName": "Fancy Apple",
          "name": "Apple",
          "taxedValue": 23.7,
          "value": 90,
        },
      },
      {
        "amount": 2,
        "ingredient": {
          "longName": "Warming Ginger",
          "name": "Ginger",
          "taxedValue": 34.7,
          "value": 109,
        },
      },
    ],
    "ingredient60": [
      {
        "amount": 4,
        "ingredient": {
          "longName": "Fancy Apple",
          "name": "Apple",
          "taxedValue": 23.7,
          "value": 90,
        },
      },
      {
        "amount": 3,
        "ingredient": {
          "longName": "Fancy Egg",
          "name": "Egg",
          "taxedValue": 38.7,
          "value": 115,
        },
      },
      {
        "amount": 3,
        "ingredient": {
          "longName": "Warming Ginger",
          "name": "Ginger",
          "taxedValue": 34.7,
          "value": 109,
        },
      },
    ],
    "ingredientPercentage": 20.7,
    "name": "PIKACHU",
    "pokedexNumber": 25,
    "previousEvolutions": 1,
    "remainingEvolutions": 1,
    "skill": {
      "RP": [
        400,
        569,
        785,
        1083,
        1496,
        2066,
        2656,
      ],
      "activations": {
        "strength": {
          "amount": [Function],
          "unit": "strength",
        },
      },
      "description": [Function],
      "image": "strength",
      "name": "Charge Strength S",
      "strengthAmounts": [
        400,
        569,
        785,
        1083,
        1496,
        2066,
        3002,
      ],
    },
    "skillPercentage": 2.1,
    "specialty": "berry",
  },
  "ribbon": 0,
  "rp": 1988,
  "saved": false,
  "shiny": false,
  "skillLevel": 2,
  "subskills": [],
  "version": 0,
}
`;

exports[`PokemonSearch > selects Pokémon correctly 2`] = `
{
  "carrySize": 22,
  "externalId": "some uuid",
  "gender": "male",
  "ingredients": [
    {
      "amount": 1,
      "ingredient": {
        "longName": "Fancy Apple",
        "name": "Apple",
        "taxedValue": 23.7,
        "value": 90,
      },
      "level": 0,
    },
    {
      "amount": 2,
      "ingredient": {
        "longName": "Fancy Apple",
        "name": "Apple",
        "taxedValue": 23.7,
        "value": 90,
      },
      "level": 30,
    },
    {
      "amount": 4,
      "ingredient": {
        "longName": "Fancy Apple",
        "name": "Apple",
        "taxedValue": 23.7,
        "value": 90,
      },
      "level": 60,
    },
  ],
  "level": 60,
  "name": "Some name",
  "nature": {
    "energy": 1,
    "exp": 1,
    "frequency": 1,
    "ingredient": 1,
    "name": "Bashful",
    "negativeModifier": "neutral",
    "positiveModifier": "neutral",
    "prettyName": "Bashful (neutral)",
    "skill": 1,
  },
  "pokemon": {
    "berry": {
      "name": "GREPA",
      "type": "electric",
      "value": 25,
    },
    "carrySize": 17,
    "displayName": "Pikachu",
    "frequency": 2700,
    "genders": {
      "female": 0.5,
      "male": 0.5,
    },
    "ingredient0": [
      {
        "amount": 1,
        "ingredient": {
          "longName": "Fancy Apple",
          "name": "Apple",
          "taxedValue": 23.7,
          "value": 90,
        },
      },
    ],
    "ingredient30": [
      {
        "amount": 2,
        "ingredient": {
          "longName": "Fancy Apple",
          "name": "Apple",
          "taxedValue": 23.7,
          "value": 90,
        },
      },
      {
        "amount": 2,
        "ingredient": {
          "longName": "Warming Ginger",
          "name": "Ginger",
          "taxedValue": 34.7,
          "value": 109,
        },
      },
    ],
    "ingredient60": [
      {
        "amount": 4,
        "ingredient": {
          "longName": "Fancy Apple",
          "name": "Apple",
          "taxedValue": 23.7,
          "value": 90,
        },
      },
      {
        "amount": 3,
        "ingredient": {
          "longName": "Fancy Egg",
          "name": "Egg",
          "taxedValue": 38.7,
          "value": 115,
        },
      },
      {
        "amount": 3,
        "ingredient": {
          "longName": "Warming Ginger",
          "name": "Ginger",
          "taxedValue": 34.7,
          "value": 109,
        },
      },
    ],
    "ingredientPercentage": 20.7,
    "name": "PIKACHU",
    "pokedexNumber": 25,
    "previousEvolutions": 1,
    "remainingEvolutions": 1,
    "skill": {
      "RP": [
        400,
        569,
        785,
        1083,
        1496,
        2066,
        2656,
      ],
      "activations": {
        "strength": {
          "amount": [Function],
          "unit": "strength",
        },
      },
      "description": [Function],
      "image": "strength",
      "name": "Charge Strength S",
      "strengthAmounts": [
        400,
        569,
        785,
        1083,
        1496,
        2066,
        3002,
      ],
    },
    "skillPercentage": 2.1,
    "specialty": "berry",
  },
  "ribbon": 0,
  "rp": 1988,
  "saved": false,
  "shiny": false,
  "skillLevel": 2,
  "subskills": [],
  "version": 0,
}
`;
